scipy/sparse/linalg/tests/test_propack.py

```diff
@@ -0,0 +1,272 @@
+import os
+import pytest
+import sys
+
+import numpy as np
+from numpy.testing import assert_allclose, assert_raises
+
+from scipy.sparse.linalg._svdp import _svdp
+from scipy.sparse import csr_matrix, csc_matrix, coo_matrix
+
+TOLS = {
+    np.float32: 1e-4,
+    np.float64: 1e-8,
+    np.complex64: 1e-4,
+    np.complex128: 1e-8,
+}
+
+_dtype_map = {
+    "single": np.float32,
+    "double": np.float64,
+    "complex8": np.complex64,
+    "complex16": np.complex128,
+}
+
+
+def is_complex_type(dtype):
+    return np.dtype(dtype).char.isupper()
+
+
+def is_32bit():
+    return sys.maxsize <= 2**32  # (usually 2**31-1 on 32-bit)
+
+
+_dtype_testing = []
+for dtype in _dtype_map:
+    if 'complex' in dtype and is_32bit():
+        # PROPACK has issues w/ complex on 32-bit; see gh-14433
+        marks = [pytest.mark.skip]
+    elif 'complex' in dtype:
+        marks = [pytest.mark.slow]  # type: ignore[list-item]
+    else:
+        marks = []
+    _dtype_testing.append(pytest.param(dtype, marks=marks))
+_dtype_testing = tuple(_dtype_testing)  # type: ignore[assignment]
+
+
+def generate_matrix(constructor, n, m, f,
+                    dtype=float, rseed=0, **kwargs):
+    """Generate a random sparse matrix"""
+    rng = np.random.RandomState(rseed)
+    if is_complex_type(dtype):
+        M = (- 5 + 10 * rng.rand(n, m)
+             - 5j + 10j * rng.rand(n, m)).astype(dtype)
+    else:
+        M = (-5 + 10 * rng.rand(n, m)).astype(dtype)
+    M[M.real > 10 * f - 5] = 0
+    return constructor(M, **kwargs)
+
+
+def assert_orthogonal(u1, u2, rtol, atol):
+    """Check that the first k rows of u1 and u2 are orthogonal"""
+    A = abs(np.dot(u1.conj().T, u2))
+    assert_allclose(A, np.eye(u1.shape[1], u2.shape[1]), rtol=rtol, atol=atol)
+
+
+def check_svdp(n, m, constructor, dtype, k, irl_mode, which, f=0.8):
+    tol = TOLS[dtype]
+
+    M = generate_matrix(np.asarray, n, m, f, dtype)
+    Msp = constructor(M)
+
+    u1, sigma1, vt1 = np.linalg.svd(M, full_matrices=False)
+    u2, sigma2, vt2, _ = _svdp(Msp, k=k, which=which, irl_mode=irl_mode,
+                               tol=tol)
+
+    # check the which
+    if which.upper() == 'SM':
+        u1 = np.roll(u1, k, 1)
+        vt1 = np.roll(vt1, k, 0)
+        sigma1 = np.roll(sigma1, k)
+
+    # check that singular values agree
+    assert_allclose(sigma1[:k], sigma2, rtol=tol, atol=tol)
+
+    # check that singular vectors are orthogonal
+    assert_orthogonal(u1, u2, rtol=tol, atol=tol)
+    assert_orthogonal(vt1.T, vt2.T, rtol=tol, atol=tol)
+
+
+@pytest.mark.parametrize('ctor', (np.array, csr_matrix, csc_matrix))
+@pytest.mark.parametrize('precision', _dtype_testing)
+@pytest.mark.parametrize('irl', (True, False))
+@pytest.mark.parametrize('which', ('LM', 'SM'))
+def test_svdp(ctor, precision, irl, which):
+    np.random.seed(0)
+    dtype = _dtype_map[precision]
+    n, m, k = 10, 20, 3
+    if which == 'SM' and not irl:
+        with assert_raises(ValueError):
```

Comment:
We should match the message (do like in `test_svds`)

Reply:
In next commit.